dotcom-rendering/src/components/TagFrontSlowMpu.tsx

```diff
@@ -0,0 +1,167 @@
+	trails,
+	adIndex,
+}: {
+	trails: Tuple<DCRFrontCard, 5>;
+	adIndex: number;
+}) => {
+	return (
+		<>
+			<UL direction="row" padBottom={true}>
+				<LI percentage="33.333%" padSides={true}>
+					<Card33Media33 trail={trails[0]} showAge={true} />
+				</LI>
+				<LI percentage="33.333%" padSides={true} showDivider={true}>
+					<Card33Media33 trail={trails[1]} showAge={true} />
+				</LI>
+				<LI percentage="33.333%" padSides={true} showDivider={true}>
+					<Card33Media33 trail={trails[2]} showAge={true} />
+				</LI>
+			</UL>
+			<UL direction="row">
+				<LI percentage="33.333%" padSides={true}>
+					<Card33Media33 trail={trails[3]} showAge={true} />
+				</LI>
+				<LI percentage="33.333%" padSides={true} showDivider={true}>
+					<Card33Media33 trail={trails[4]} showAge={true} />
+				</LI>
+				<LI percentage="33.333%" padSides={true} showDivider={true}>
+					<Hide until="tablet">
+						<AdSlot position="inline" index={adIndex} />
+					</Hide>
+				</LI>
+			</UL>
+		</>
+	);
+};
+
+export const SevenCards = ({
+	trails,
+	adIndex,
+}: {
+	trails: [...Tuple<DCRFrontCard, 7>, ...DCRFrontCard[]];
+	adIndex: number;
+}) => {
+	return (
+		<>
+			<UL direction="row" padBottom={true}>
+				<LI percentage="50%" padSides={true}>
+					<Card50Media50 trail={trails[0]} showAge={true} />
+				</LI>
+				<LI percentage="50%" padSides={true} showDivider={true}>
+					<Card50Media50 trail={trails[1]} showAge={true} />
+				</LI>
+			</UL>
+			<UL direction="row" padBottom={true}>
+				<LI percentage="33.333%" padSides={true}>
+					<Card33Media33 trail={trails[2]} showAge={true} />
+				</LI>
+				<LI percentage="33.333%" padSides={true} showDivider={true}>
+					<Card33Media33 trail={trails[3]} showAge={true} />
+				</LI>
+				<LI percentage="33.333%" padSides={true} showDivider={true}>
+					<Card33Media33 trail={trails[4]} showAge={true} />
+				</LI>
+			</UL>
+			<UL direction="row">
+				<LI percentage="33.333%" padSides={true}>
+					<Card33Media33 trail={trails[5]} showAge={true} />
+				</LI>
+				<LI percentage="33.333%" padSides={true} showDivider={true}>
+					<Card33Media33 trail={trails[6]} showAge={true} />
+				</LI>
+				<LI percentage="33.333%" padSides={true} showDivider={true}>
+					<Hide until="tablet">
+						<AdSlot position="inline" index={adIndex} />
+					</Hide>
+				</LI>
+			</UL>
+		</>
+	);
+};
+
+export const TagFrontSlowMpu = ({ trails, adIndex }: Props) => {
+	if (isTuple(trails, 2)) {
+		return <TwoCard trails={trails} adIndex={adIndex} />;
+	} else if (isTuple(trails, 4)) {
+		return <FourCard trails={trails} adIndex={adIndex} />;
+	} else if (isTuple(trails, 5)) {
+		return <FiveCard trails={trails} adIndex={adIndex} />;
+	} else if (isTuple(trails, 7)) {
+		return <SevenCards trails={trails} adIndex={adIndex} />;
+	} else {
+		// The above if statements cover all the types that trails could be, but ESLint gets upset that "not all code paths return a value"
+		return <></>;
+	}
+};
```

Comment:
The beauty of tuples is really that you can have exhaustive checks. Here the value of `trails.length` is narrowed to `2 | 4 | 6 | 9`

```ts
export const TagFrontFastMpu = ({ trails, adIndex }: Props) => {
	switch (trails.length) {
		case 2:
			return <TwoCard trails={trails} adIndex={adIndex} />;
		case 4:
			return <FourCard trails={trails} adIndex={adIndex} />;
		case 6:
			return <SixCard trails={trails} adIndex={adIndex} />;
		case 9:
			return <NineCard trails={trails} adIndex={adIndex} />;
	}
};
```

Reply:
Amazing ! Now improved